File: old-projects/vector-defence-silverlight-port/src/Towers/LaserTower.ts
import { Point } from "../Point";
import { Game } from "../Game";
import { Util } from "../Util";
import { LaserRay } from "../Particles/LaserRay";
import { Monster } from "../Monsters/Monster";
import { Tower } from "./Tower";

export class LaserTower extends Tower {
  private readonly ray = new LaserRay();
  private trackedMonster: Monster | null = null;
  private gunTip = new Point(0, 0);
  private gunAngle = 0;

  constructor(x: number, y: number, addToGame = true) {
    super(x, y, addToGame);
    this.cost = 30;
    this.range = 100;
    this.millisecondsBetweenFiring = 1500;
    this.gunAngle = Util.randomInRange(-Math.PI, Math.PI);
    this.turnGun();
  }

  override removeTowerFromCanvas(): void {
    super.removeTowerFromCanvas();
    this.ray.markedForRemoval = true;
  }

  override render(context: CanvasRenderingContext2D): void {
    this.turnGun();
    context.save();
    context.fillStyle = "#000";
    context.beginPath();
    context.arc(this.x, this.y, 12.5, 0, Math.PI * 2);
    context.fill();
    context.translate(this.x, this.y);
    context.rotate(this.gunAngle);
    context.fillStyle = "#00ffff";
    context.strokeStyle = "#fff";
    context.lineWidth = 1.5;
    context.beginPath();
    context.moveTo(-10, 5);
    context.lineTo(-2, 4);
    context.lineTo(10, 0);
    context.lineTo(-2, -4);
    context.lineTo(-10, -5);
    context.closePath();
    context.fill();
    context.stroke();
    context.restore();
  }

  protected override animate(): void {
    if (this.trackedMonster === null || this.trackedMonster.markedForRemoval) {
      this.trackedMonster = this.getClosestMonsterInRange();
    }

    if (this.trackedMonster !== null) {
      const closeEnough = Util.isWithinDistance(this.trackedMonster.x, this.trackedMonster.y, this.x, this.y, this.range);
      if (!closeEnough || this.trackedMonster.markedForRemoval) {
        this.trackedMonster = null;
      } else {
        const sourcePoint = this.gunTip.clone();
        if (this.ray.isOn) {
          const target = this.turnGunToTrackedMonster();
          this.ray.turnTo(sourcePoint, target);
        }
        if (this.canFireNow) {
          const target = this.turnGunToTrackedMonster();
          this.ray.fireAt(sourcePoint, target);
          this.timeFiring();
        }
      }
    }
  }

  protected override upgradeWork(newLevel: number): void {
    super.upgradeWork(newLevel);
    this.ray.setLevel(newLevel);
  }

  private turnGunToTrackedMonster(): Point {
    const target = new Point(this.trackedMonster?.x ?? this.x, this.trackedMonster?.y ?? this.y);
    this.gunAngle = Util.calculateAngle(this.x, this.y, target.x, target.y);
    this.turnGun();
    return target;
  }

  private turnGun(): void {
    this.gunTip = new Point(
      this.x + (9 * Math.cos(this.gunAngle)),
      this.y + (9 * Math.sin(this.gunAngle)),
    );
  }
}

File: silverlight-ports/vector-defence/original-csharp-source/Towers/LaserTower.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Shapes;
using VectorDefenceSL.Monsters;
using VectorDefenceSL.Particles;

namespace VectorDefenceSL.Towers
{
    public class LaserTower : Tower
    {
        private readonly LaserRay ray = new LaserRay();
        private readonly RotateTransform transform = new RotateTransform();
        private readonly Polygon gunPolygon = new Polygon
        {
            IsHitTestVisible = false,
            StrokeThickness = 1.5,
            Stroke = new SolidColorBrush(Colors.White),
            Fill = new SolidColorBrush(Colors.Cyan),
            Points =
            {
                new Point(-10, 5),
                new Point(-2, 4),
                new Point(10, 0),
                new Point(-2, -4),
                new Point(-10, -5)
            }
        };

        private readonly Ellipse ellipse = new Ellipse
        {
            Width = 25,
            Height = 25,
            Fill = new SolidColorBrush(Colors.Black)
        };

        private Monster trackedMonster;
        private Point gunTip;
        private double gunAngle;

        public LaserTower(double x, double y) : base(x, y)
        {
            Cost = 30;
            Range = 100;
            MillisecondsBetweenFiring = 1500;
            gunPolygon.RenderTransform = transform;
            ellipse.MouseLeftButtonDown += HandleTowerSelection;
            Container.Children.Add(ellipse);
            Container.Children.Add(gunPolygon);
        }

        public LaserTower(Button btn) : base(btn, TowerBuildingMode.LaserTower)
        {
            gunPolygon.RenderTransform = transform;
            Container.Children.Add(gunPolygon);
        }

        public override void RemoveTowerFromCanvas()
        {
            base.RemoveTowerFromCanvas();
            Container.Children.Remove(ellipse);
            Container.Children.Remove(gunPolygon);
            ray.RemoveFromCanvas();
        }

        public override sealed void Render()
        {
            gunAngle = Util.RandomInRange(-Math.PI, Math.PI);
            TurnGun();
            gunPolygon.SetCanvasPosition(X, Y);
            ellipse.SetCanvasPosition(X - (ellipse.Width / 2), Y - (ellipse.Height / 2));
        }

        protected override void Animate()
        {
            if (trackedMonster == null || trackedMonster.MarkedForRemoval)
            {
                trackedMonster = GetClosestMonsterInRange();
            }
            
            if (trackedMonster != null)
            {
                bool closeEnough = Util.IsWithinDistance(trackedMonster.X, trackedMonster.Y, X, Y, Range);
                if (!closeEnough || trackedMonster.MarkedForRemoval)
                {
                    trackedMonster = null;
                }
                else
                {
                    var sourcePoint = new Point(gunTip.X, gunTip.Y);
                    if (ray.IsOn)
                    {
                        Point target = TurnGunToTrackedMonster();
                        ray.TurnTo(sourcePoint, target);
                    }

                    if (CanFireNow)
                    {
                        Point target = TurnGunToTrackedMonster();
                        ray.FireAt(sourcePoint, target);
                        TimeFiring();
                    }
                }
            }
        }

        protected override void AnimateToolboxRendering()
        {
            gunAngle += ToolbarAnimationRotationStep * Game.CurrentFrameMultiplier;
            TurnGun();            
        }

        protected override void UpgradeWork(int newLevel)
        {
            base.UpgradeWork(newLevel);
            ray.SetLevel(newLevel);
        }

        private Point TurnGunToTrackedMonster()
        {
            var target = new Point(trackedMonster.X, trackedMonster.Y);
            gunAngle = Util.CalculateAngle(X, Y, target.X, target.Y);
            TurnGun();
            return target;
        }

        private void TurnGun()
        {
            transform.Angle = Util.GetAngleInDegrees(gunAngle);
            gunTip.X = X + (9 * Math.Cos(gunAngle));
            gunTip.Y = Y + (9 * Math.Sin(gunAngle));
        }
    }
}